edgelesssys/constellation
# bazel: build container images

### Proposed change(s)

- bazel: convert all container images to Bazel
- hack: add oci-pin tool
  - tool can generate pinning artifacts for container images during a `bazel build`:
    - Go source files that contain constants that reference a pinned image
    - sh256sums files
- cli: use Bazel container images
- A new command `bazel run //:push` that builds + pushes all container images so they are up to date.
  - The command uses the exact same images also pinned by the CLI.
  - `//:devbuild` also generates a file called `container_sums.sha256` that contains all pinned container images for easier checks
- e2e tests and release pipeline now upload container images using bazel *just-in-time* (whenever the CLI is being built).
- A lot of cleanup (apko, raw Dockerfiles, workflows)  



### How to test

Create a cluster and validate that

- the pinned images are uploaded
- the pinned images are deployed in the cluster
- everything still works as expected

```shell-session
bazel run //:devbuild
bazel run //:push
# create a constellation as usual
# ...

# list all container images used in the cluster
kubectl get pods --all-namespaces -o jsonpath="{.items[*].spec.containers[*].image}" |\
    tr -s '[[:space:]]' '\n' |\
    sort |\
    uniq -c

# compare to what is pinned
cat container_sums.sha256
```

### Open tasks (for future PRs)

This PR got quite large so I decided to push more features into separate PRs:

- Move all external container images we depend on to Bazel
- Allow developers to push container images into their own registry
- Add airgapping config settings + feature to `constellation-config.yaml`

### Checklist
